okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealCall.kt

```diff
@@ -118,6 +118,8 @@ class RealCall(
   @Volatile private var exchange: Exchange? = null
   internal val plansToCancel = CopyOnWriteArrayList<RoutePlanner.Plan>()
 
+  @Volatile private var onCancels: List<() -> Unit> = listOf()
```

Comment:
why volatile?

Reply:
This class handles concurrency pretty well, with this I didn't really put enough thought into how to do this right.  but race conditions between people registering listeners and a cancelation from another thread probably deserves some consideration, or documentation of the exact rules to exclude race conditions.

I'm going to make a stab at it, but please review for correctness and roughly optimal behaviour in typical cases...